lib/postgrex/stream.ex

```diff
@@ -0,0 +1,70 @@
+defmodule Postgrex.StreamResult do
+  defstruct [:columns, :row]
+end
+
+defmodule Postgrex.Stream do
+  defstruct [:conn, :max_rows, :options, :params, :portal, :query, :rows, state: :ready]
+
+  defimpl DBConnection.Query do
+    alias Postgrex.Stream
+    alias DBConnection.Query
+
+    def parse( query, opts ) do
+      %Stream{ query: Query.parse( query, opts ) }
+    end
+
+    def describe( %Stream{ query: query } = stream, opts ) do
+      %Stream{ stream | query: Query.describe( query, opts ) }
+    end
+
+    def encode( %Stream{ query: query } = stream , params, opts ) do
+      %Stream{ stream | query: Query.encode( query, params, opts ) }
+    end
+
+    def decode( %Stream{ query: query } = stream , result, opts ) do
+      %{ rows: rows } = Query.decode( query, result, opts )
+      %Stream{ stream | rows: rows }
+    end
+  end
+
+  defimpl Enumerable do
+    alias Postgrex.Stream
+
+    def reduce(      _, {:halt,    acc},   _ ), do: {:halted,    acc}
+    def reduce( stream, {:suspend, acc}, fun ), do: {:suspended, acc, &reduce( stream, &1, fun ) }
```

Comment:
Considering using `Stream.resource/3` and closing the portal in the after fun. This should tidy up the reduce code considerably. Also I think we should use `Postgrex.execute!` so that we raise on an error and only pass `Postgrex.Result.t` as elements of the stream.

`File.Stream` and `IO.Stream` are good examples for both of these points.